lib/ClangImporter/ClangImporter.cpp

```diff
@@ -430,49 +430,54 @@ static void
 getNormalInvocationArguments(std::vector<std::string> &invocationArgStrs,
                              ASTContext &ctx,
                              const ClangImporterOptions &importerOpts) {
-  const llvm::Triple &triple = ctx.LangOpts.Target;
+  const auto &LangOpts = ctx.LangOpts;
+  const llvm::Triple &triple = LangOpts.Target;
   SearchPathOptions &searchPathOpts = ctx.SearchPathOpts;
 
   auto languageVersion = ctx.LangOpts.EffectiveLanguageVersion;
 
-  if (llvm::sys::path::extension(importerOpts.BridgingHeader).endswith(
-        PCH_EXTENSION)) {
-    invocationArgStrs.insert(
-      invocationArgStrs.end(),
-        { "-include-pch", importerOpts.BridgingHeader }
-    );
+  if (llvm::sys::path::extension(importerOpts.BridgingHeader)
+          .endswith(PCH_EXTENSION)) {
+    invocationArgStrs.insert(invocationArgStrs.end(), {
+        "-include-pch", importerOpts.BridgingHeader
+    });
   }
 
   // Construct the invocation arguments for the current target.
   // Add target-independent options first.
-  invocationArgStrs.insert(
-      invocationArgStrs.end(),
-      {
+  invocationArgStrs.insert(invocationArgStrs.end(), {
+      // Enable modules
+      "-fmodules",
+      "-Werror=non-modular-include-in-framework-module",
+      "-Xclang", "-fmodule-feature", "-Xclang", "swift",
 
-          // Enable modules
-          "-fmodules",
-          "-Werror=non-modular-include-in-framework-module",
-          "-Xclang", "-fmodule-feature", "-Xclang", "swift",
+      // Don't emit LLVM IR.
+      "-fsyntax-only",
 
-          // Don't emit LLVM IR.
-          "-fsyntax-only",
+      // Enable block support.
+      "-fblocks",
 
-          // Enable block support.
-          "-fblocks",
+      languageVersion.preprocessorDefinition("__swift__", {10000, 100, 1}),
 
-          languageVersion.preprocessorDefinition("__swift__", {10000, 100, 1}),
+      "-fretain-comments-from-system-headers",
 
-          "-fretain-comments-from-system-headers",
+      SHIMS_INCLUDE_FLAG, searchPathOpts.RuntimeResourcePath,
+  });
 
-          SHIMS_INCLUDE_FLAG, searchPathOpts.RuntimeResourcePath,
-      });
+  if (LangOpts.EnableObjCInterop) {
+    invocationArgStrs.insert(invocationArgStrs.end(),
+                             {"-x", "objective-c", "-std=gnu11", "-fobjc-arc"});
+    // TODO: Investigate whether 7.0 is a suitable default version.
+    if (!triple.isOSDarwin())
+      invocationArgStrs.insert(invocationArgStrs.end(),
+                               {"-fobjc-runtime=ios-7.0"});
```

Comment:
This will let us make some progress, but eventually you'll need to decide which Objective-C runtime(s) to target outside of Darwin. For example, Clang supports GNUstep, and one could imagine surfacing Clang's understanding of the Objective-C runtime so that Swift could build on top of it (rather than duplicating it).

Reply:
I don't have much time to work on it, but I'm happy to provide advice to anyone who wants to add support for the GNUstep runtime (and add any missing runtime features that Swift needs).

Reply:
Until we get GNUStep to the point where the object layout works in a compatible manner (or change the swift object layout), I think that not permitting control over the runtime is quite reasonable.

Reply:
Are Swift's requirements documented anywhere?

Reply:
I don't think so. Off the top of my head two of the biggest requirements are:
* Swift wants to generate ObjC class structures (at compile time and at runtime) with additional Swift data prepended and appended. 
* Swift wants control of a bit in the ObjC class structure in order to mark ObjC classes that are also Swift classes.

Reply:
The second requirement is easy - we have a flags field and can reserve a bit for Swift to use.

The first is a bit harder because the clang generates these structures without providing hooks to add fields.  The structure is [defined here]( https://github.com/gnustep/libobjc2/blob/master/class.h#L39).

This will change in the next version of the runtime, and the current prototype for the new ABI would make this more difficult, because it puts all of the class structures in a section and the loader in the runtime expects them to be contiguous in memory, so adding other fields would mean that the runtime would need to be aware of the structure.

I'm also rearranging some of the fields to split the structure into ones that are likely to be modified at run time and ones that aren't, based on feedback from Microsoft that the runtime generates more CoW faults at start than they'd like - I'm not sure where the Swift fields would like to go, but if it's on the immutable part then that's probably easier, because that will be reachable via a pointer to the start from the mutable part, and that can easily be a pointer to the middle of a larger structure.